File: c++/life.cpp
#include <iostream>
#include <set>
#include <unordered_set>
#include <vector>
#include <algorithm>
#include <utility>
#include <thread>
#include <chrono>

using namespace std;

namespace life {

struct Position {
  int x, y;
  // if we put Position in a set, define an order:
  bool operator<(const Position & b) const {
    if (x == b.x) {
      return y < b.y;
    }
    return x < b.x;
  }
  // if we use unordered_set, define equality
  bool operator==(const Position & b) const {
    return x == b.x && y == b.y;
  }
  Position(int x, int y) : x(x), y(y) {}
  Position(const Position & p) : x(p.x), y(p.y) {}
};
struct CompareX {
  bool operator()(const Position & a, const Position & b) const { return a.x < b.x; }
};
struct CompareY {
  bool operator()(const Position & a, const Position & b) const { return a.y < b.y; }
};

struct Hash {
  size_t operator()(const Position &position) const {
    return position.x * 65521 + position.y;
  }
};

// for debugging
ostream & operator<<(ostream &ostr, const Position & p) {
  ostr << "(" << p.x << ", " << p.y << ")";
  return ostr;
}

enum How { Live, Die, Unchanged };

ostream & operator<<(ostream &ostr, How how) {
  switch (how) {
  case Live:
    cout << "Live";
    break;
  case Die:
    cout << "Die";
    break;
  case Unchanged:
    cout << "Unchanged";
    break;
  }
  return ostr;
}

struct Change {
  How how;
  Position position;
  Change(How how, const Position & pos) : how(how), position(pos) {}
};

const Change DoNothing(Unchanged, Position(0, 0));

// more debugging
ostream & operator<<(ostream &ostr, const Change & c) {
  ostr << "how=" << c.how << ", position=" << c.position;
  return ostr;
}

// Hashed sets are about 12% faster for me
typedef unordered_set<Position, Hash> Positions;
//typedef set<Position> Positions;

struct Board {
  const Positions alive;
  const vector<Change> updates;

  Board(const Positions & alive,
	const vector<Change> & updates)
    : alive(alive),
      updates(updates)
  {
  }
};

// debugging
ostream & operator<<(ostream &ostr, const Board & b) {
  ostr << "alive=";
  string sep = "";
  for (auto each : b.alive) {
    ostr << sep << each;
    sep = ", ";
  }
  ostr << "updates=";
  sep = "";
  for (auto each : b.updates) {
    ostr << sep << each;
    sep = ", ";
  }
  return ostr;
 }

Position p(int x, int y) { return Position(x, y); }
const Position r_pentomino[] = {
  p(0, 0), p(0, 1), p(1, 1), p(-1, 0), p(0, -1)
};

void clearScreen() {
  // clear screen
  cout << char(27) << "[2J";
  // move to top-left corner
  cout << char(27) << "[;H";
}

pair<Position, Position> boundBox(const Board & board) {
  const Positions & lst = board.alive;
  if (lst.empty()) {
    return make_pair(Position(0,0), Position(1,1));
  }
  auto mmx = minmax_element(lst.begin(), lst.end(), CompareX());
  auto mmy = minmax_element(lst.begin(), lst.end(), CompareY());
  return make_pair(Position(mmx.first->x, mmy.first->y),
		   Position(mmx.second->x, mmy.second->y));
}

void printBoard(const Board & board) {
  auto bb = boundBox(board);
  auto min = bb.first;
  auto max = bb.second;
  for (auto y = max.y; y >= min.y; y--) {
    for(auto x = min.x; x <= max.x; x++) {
      cout << (board.alive.count(Position(x, y)) ? '@' : ' ');
    }
    cout << endl;
  }
}

// apply the Die/Live changes to the live set
Positions applyUpdates(const Positions & alive,
			   const vector<Change> & updates) {
  // use the same cheat in Java, but it's much harder to not
  // do it this way in c++
  Positions result(alive);
  for (auto & change : updates) {
    if (change.how == Die) {
      result.erase(change.position);
    } else if (change.how == Live) {
      result.insert(change.position);
    }
  }
  return result;
}

// harder to type, but easier to understand I think:
const int OFFSETS[][2] = {
  {-1, 1},  {0,  1}, {1,  1},
  {-1, 0},           {1,  0},
  {-1, -1}, {0, -1}, {1, -1}
};
const size_t OFFSET_COUNT = sizeof(OFFSETS)/sizeof(OFFSETS[0]);

// generate the eight neighbor positions of a given position
vector<Position> eight(const Position & position) {
  vector<Position> result;
  result.reserve(OFFSET_COUNT);
  for (size_t i = 0; i < OFFSET_COUNT; i++) {
    result.push_back(Position(position.x + OFFSETS[i][0],
			      position.y + OFFSETS[i][1]));
  }
  return result;
}

// generate the set of all affected neighbors for a ChangeSet
Positions neighbors(const vector<Change> & changes) {
  Positions result;
  result.reserve(changes.size() * 8);
  for (auto & change : changes) {
    for (auto & pos : eight(change.position)) {
      result.insert(pos);
    }
  }
  return result;
}

// compute the state change for the next generation at a given position
Change change(const Positions & alive, const Position & pos) {
  int liveCount = 0;
  for (auto & n : eight(pos)) {
    liveCount += alive.count(n);
  }
  if (liveCount == 2) {
    return DoNothing;
  }
  if (alive.count(pos)) {
    if (liveCount != 3) {
      return Change(Die, pos);
    }
  } else {
    if (liveCount == 3) {
      return Change(Live, pos);
    }
  }
  return DoNothing;
}

// get the set of changes to apply to the next generation for a set of points
vector<Change> computeChanges(const Positions & alive,
			      const Positions & affected) {
  // again, cheating with a local mutable type
  vector<Change> result;
  for (auto & pos : affected) {
    Change c = change(alive, pos);
    if (c.how != Unchanged) {
      result.push_back(c);
    }
  }
  return result;
}

// compute a new board from the old board
unique_ptr<Board> nextGeneration(const Board & board) {
  // this came out nicer than I was expecting
  auto alive = applyUpdates(board.alive, board.updates);
  auto affected = neighbors(board.updates);
  auto updates = computeChanges(alive, affected);
  return unique_ptr<Board>(new Board(alive, updates));
}

} // namespace life

using namespace life;

int main(int argc, char **argv) {
  vector<Change> start;
  for (size_t i = 0; i < sizeof(r_pentomino) / sizeof(r_pentomino[0]); i++) {
    start.push_back(Change(Live, r_pentomino[i]));
  }
  unique_ptr<const Board> board(new Board(Positions(), start));
  const auto generations = 1000;
  const auto showWork = false;
  const auto times = 5;
  const auto human_speed = chrono::milliseconds(1000 / 30);
  for (int time = 0; time < times; time++) {
    auto start = std::chrono::system_clock::now();
    for(int i = 0; i < generations; i++) {
      board = nextGeneration(*board);
      if (showWork) {
	clearScreen();
	printBoard(*board);
	this_thread::sleep_for(human_speed);
      }
    }
    auto end = std::chrono::system_clock::now();
    auto diff_us = chrono::duration_cast<chrono::microseconds>(end - start);
    auto diff_s = diff_us.count() / 1000. / 1000.;
    cout << (generations / diff_s) << " generations / sec" << endl;
  }
  return 0;
}

File: nim/life.nim
import std/[sets, sequtils, os, sugar, options, times]

type
  Coord = tuple
    x, y: int

proc coord(p : (int, int)): Coord =
  let (x, y) = p
  return (x: x, y: y)

type
  Change = tuple
    alive: bool
    coord: Coord

proc kill(coord: Coord): Change =
  return (alive: false, coord: coord)

proc live(coord: Coord): Change =
  return (alive: true, coord: coord)

type
  Board = tuple
    alive: HashSet[Coord]
    updates: seq[Change]

proc clearScreen() =
  # clear screen
  stdout.write "\e[2J"
  # move to top-left corner
  stdout.write "\e[;H"
  stdout.flushFile()

proc lowerLeft(a: Coord, b: Coord): Coord =
  return (min(a.x, b.x), min(a.y, b.y))

proc upperRight(a: Coord, b: Coord): Coord =
  return (max(a.x, b.x), max(a.y, b.y))

proc boundBox(board: Board): (Coord, Coord)  =
  let lst = board.alive.toSeq()
  if lst.len() == 0:
    return ((0, 0), (1, 1))
  let ll = foldl(lst, lowerLeft(a, b))
  let ur = foldl(lst, upperRight(a, b))
  return (ll, ur)

proc printBoard(board: Board) =
  let human_speed = int(1000 / 30)
  clearScreen()
  let (min, max) = boundBox(board)
  for y in countdown(max.y, min.y):
    for x in min.x .. max.x:
      if board.alive.contains((x, y)):
        stdout.write('@')
      else:
        stdout.write(' ')
    stdout.write('\n')
  stdout.flushFile()
  sleep(human_speed)

# apply the kill/resurection set to the live set
proc applyUpdates(alive : SomeSet[Coord], updates: seq[Change]): SomeSet[Coord] =
  let toLive = updates.filterIt(it.alive).mapIt(it.coord).toHashSet()
  let toDie = updates.filterIt(not it.alive).mapIt(it.coord).toHashSet()
  return (toLive + alive) - toDie

let OFFSETS* = collect(newSeq):
    for x in -1..1:
      for y in -1..1:
        if x != 0 or y != 0: (x, y)

iterator eight(coord: Coord) : Coord {.inline.} =
  for (xoff, yoff) in OFFSETS:
    yield (coord.x + xoff, coord.y + yoff)

iterator neighborsOfCoords(coords: seq[Coord]) : Coord {.inline.} =
  for coord in coords:
    for neighbor in eight(coord):
      yield neighbor

proc affected(changes: seq[Change]): SomeSet[Coord] =
  return changes.toSeq().mapIt(it.coord).neighborsOfCoords().toSeq().toHashSet()

# compute the state change for the next generation at a given coord
proc computeChange(alive: SomeSet[Coord], coord: Coord): Option[Change] =
  let liveCount = eight(coord).countIt(alive.contains(it))
  if liveCount == 2:
    return none(Change)
  if coord in alive:
    if liveCount != 3:
      return some(kill(coord))
  else:
    if liveCount == 3:
      return some(live(coord))
  return none(Change)

# get the set of changes to apply to the next generation for a set of points
proc computeChanges(alive: SomeSet[Coord], affected: seq[Coord]): SomeSet[Change] =
  return affected.mapIt(computeChange(alive, it)).filterIt(it.isSome).mapIt(it.get()).toHashSet()

# compute a new board from the old board
proc nextGeneration(board: Board): Board =
  let alive = applyUpdates(board.alive, board.updates)
  let affected = affected(board.updates)
  let updates = computeChanges(alive, affected.toSeq())
  return (alive, updates.toSeq())
  
proc start(alive: SomeSet[Coord]): Board =
  # turn on the start nodes
  let birth = alive.mapIt(live(it)).toHashSet()
  # mark the surrounding nodes as off for the complete affected set
  let death = (neighborsOfCoords(alive.toSeq()).toSeq().toHashSet() - alive).mapIt(kill(it)).toHashSet()
  let empty = initHashSet[Coord]()
  let updates = birth + death
  return (empty, updates.toSeq())

proc run1(board : Board, generations: int, showWork: bool) =
  if generations == 0:
    return
  let next = nextGeneration(board)
  if showWork:
    printBoard(next)
  run1(next, generations - 1, showWork)

proc main() =
  let r_pentomino = toHashSet([(0, 0), (0, 1), (1, 1), (-1, 0), (0, -1)])
  let generations = 1000
  let showWork = false
  let times = (if showWork: 1 else: 5)
  for t in 1..times:
    let now = cpuTime()
    run1(start(r_pentomino), generations, showWork)
    let diff = cputime() - now;
    echo generations.float / diff, " generations / sec"

main()
